File: Remove Nth Node From End of List/Remove Nth Node From End of List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {

    // identify nth node
    // prev.next = nthNode.next;

    // GOAL: remove nth node
    function recurse(prev, curr) {
        
        let currPos;
        
        if (curr.next == null) { // base case =  we find the tail
            // n == 1
            currPos = 1; // currPos = 1
        }
        else { // recursive call
            currPos = 1 + recurse(curr, curr.next);
        }

        

        // removal operation
        // prev.next = curr.next;
        if ( currPos == n && curr == head) { // if removing head
            head = head.next;
        }
        else if ( currPos == n ) {
            prev.next = curr.next;
        }
        // return
        return currPos;
    }

    recurse(null, head);
    return head;
};
/*
n = 2
1 -> 2

prev = null
curr = 1

null.next = 2
*/

File: Remove Nth Node From End of List/Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        hat = ListNode(-1, head)
        leader, follower = hat, hat

        # give leader headstart
        for _ in range(n): leader = leader.next

        # iterate both pointers together until leader reaches the tail
        while leader.next != None:
            leader = leader.next
            follower = follower.next

        # splice out LL[-n]
        follower.next = follower.next.next

        return hat.next